File: backend/src/main/scala/domains/user/http/api/ListUsers.scala
package domains.user.http.api

import cats.effect.IO
import domains.auth.http.SiteManagerApi
import domains.auth.objects.SiteManagerUser
import domains.user.http.codec.UserHttpCodecs.given
import domains.user.objects.request.{UserListRequest, UserSearchQuery}
import domains.user.objects.response.UserListResponse
import domains.user.table.user.UserTable
import io.circe.Encoder
import org.http4s.{Method, Request, Status}
import shared.http.utils.PageRequestQuerySupport
import shared.http.{ApiPath, PathParams}

import java.sql.Connection

object ListUsers extends SiteManagerApi[UserListRequest, UserListResponse]:

  override val method: Method = Method.GET
  override val path: ApiPath = ApiPath("/api/users")
  override val successStatus: Status = Status.Ok
  override protected val outputEncoder: Encoder[UserListResponse] = summon[Encoder[UserListResponse]]

  override def decode(request: Request[IO], pathParams: PathParams): IO[UserListRequest] =
    val _ = pathParams
    IO.pure(
      UserListRequest(
        query = request.uri.query.params.get("q").flatMap(rawQuery => UserSearchQuery.parse(rawQuery).toOption),
        pageRequest = PageRequestQuerySupport.parsePageRequest(request.uri.query.params)
      )
    )

  override def plan(connection: Connection, actor: SiteManagerUser, request: UserListRequest): IO[UserListResponse] =
    UserTable.listUsers(connection, actor, request)

File: frontend/src/apis/user/ListUsers.ts
import type { APIWithSessionMessage } from '@/system/api/api-message'
import type { UserListRequest } from '@/objects/user/request/UserListRequest'
import type { UserListResponse } from '@/objects/user/response/UserListResponse'

/** 查询受管用户列表；输入过滤和分页，输出包含邮箱/权限的管理分页响应。 */
export class ListUsers implements APIWithSessionMessage<UserListResponse> {
  declare readonly responseType?: UserListResponse
  readonly method = 'GET'
  readonly apiPath: string

  constructor(request: UserListRequest) {
    const params = new URLSearchParams()
    if (request.query !== null && request.query.trim()) {
      params.set('q', request.query)
    }
    params.set('page', String(request.pageRequest.page))
    params.set('pageSize', String(request.pageRequest.pageSize))
    this.apiPath = `users?${params.toString()}`
  }

  body(): undefined {
    return undefined
  }
}
